Add documentation on diagnosing .NET Core performance issues in containers (#20511)
dotnet/docs: docs/core/diagnostics/diagnostics-in-containers.md

---
title: Collect diagnostics in containers
description: In this article, you'll learn how .NET Core diagnostics tools can be used in Docker containers.
ms.date: 09/01/2020
---

# Collect diagnostics in containers

> **This article applies to: ✔️** .NET Core 2.1 SDK and later versions

Reviewer: ```suggestion
**This article applies to: ✔️** .NET Core 3.0 SDK and later versions
```

While the tools can be installed using a 2.1 SDK, they aren't able to work with a 2.1 application.  I saw you called this out later, but the docs page will have 2.1 specifically called out in bold font at the top of the page and users might miss your caveat statement.  I'm not 100% on what the requirements are for this statement in the docs though, so I might be wrong in suggesting this change.  @adegeo can you chime in here?

Reviewer 2: I would go with 3.1 as 3.0 is out of support.

```suggestion
**This article applies to: ✔️** .NET Core 3.1 SDK and later versions
```

Author: Yea, I wasn't sure what to put for version here because some of the tools I mention (createdump) work on 2.1, others (dotnet-dump, dotnet-trace, dotnet-counters) work with 3.0, and one (dotnet-gcdump) requires 3.1. My solution was to use the earliest version and try to call out what requires a later one.

@adegeo - do we have a usual practice when writing conceptual docs like this that apply differently to different versions?

Reviewer 2: I would use something similar to the "this article applies to" syntax but reword it for "this command" or "this tool" in the applicable section.

```markdown
## Using `PerfCollect` in a container

**This tool applies to: ✔️** .NET Core 3.1 SDK and later versions.

## Using `CreateDump` in a container

**This tool applies to: ✔️** .NET Core 2.1 SDK and later versions.

blah blah blah blah text text paragraph paragraph ...

blah blah blah blah text text paragraph paragraph ...
```

Author: Makes sense. I've removed the "This article applies to..." and added more specific "This tool applies to..." annotations for the different tools.

The same diagnostics tools that are useful for diagnosing .NET Core issues in other scenarios also work in Docker containers. However, some of the tools require special steps to work in a container. This article covers how tools for gathering performance traces and collecting dumps can be used in Docker containers.

> ## Using .NET Core global CLI tools in a container

Reviewer 2: This header is very large, it will wrap on the published page.

Author: I've updated it to `## Using .NET Core CLI tools in a container` which is a bit shorter. Let me know if you think I need to trim more.

The .NET Core global CLI diagnostic tools ([`dotnet-counters`](dotnet-counters.md), [`dotnet-dump`](dotnet-dump.md), [`dotnet-gcdump`](dotnet-gcdump.md), and [`dotnet-trace`](dotnet-trace.md)) are designed to work in a wide variety of environments and should all work directly in Docker containers. Because of this, these tools are the preferred method of collecting diagnostic information for .NET Core scenarios targeting .NET Core 3.0 or above (or 3.1 or above in the case of `dotnet-gcdump`) in containers.

> The only complicating factor of using these tools in a container is that they are installed with the .NET Core SDK and many Docker containers run without the .NET Core SDK present. One easy solution to this problem is to install the tools in the initial Docker image. The tools don't need the .NET Core SDK to run, only to be installed. Therefore, it's possible to create a Dockerfile that installs the tools in a build stage (where the .NET Core SDK is present) and then copies the binaries into the final image. The only downside to this approach is increased Docker image size (each tool is roughly 100 KB, give or take, depending on the specific tool and OS).

Reviewer 3: dotnet-dump in particular is a lot larger than 100kb. I'd omit mentioning numbers here

Author: I'll pull out the specific numbers.

> 

Author: Good idea; added.

```dockerfile
# In build stage
# Install desired .NET CLI diagnostics tools
RUN dotnet tool install --tool-path /tools dotnet-trace
RUN dotnet tool install --tool-path /tools dotnet-counters
RUN dotnet tool install --tool-path /tools dotnet-dump

...

# In final stage
# Copy diagnostics tools
WORKDIR /tools
COPY --from=build /tools .
```

> Alternatively, the .NET Core SDK can be installed in a container when needed in order to install the CLI tools. Be aware that installing the .NET Core SDK will have the side-effect of reinstalling the .NET Core runtime. So be sure to install the version of the SDK that matches the runtime present in the container.

Reviewer 4: @josalem I thought we were adding direct download links so we could also fetch the tools using `wget`/`curl`

Reviewer: We're close, but we don't have them yet.  It isn't worth documenting until they're up and we've fleshed out the instructions for installation.  This doc should definitely be updated once we have that.

### Using .NET Core global CLI tools in a sidecar container

If you would like to use .NET Core global CLI diagnostic tools to diagnose processes in a different container, bear the following additional requirements in mind:

> 1. The containers must, of course, [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid) (so that tools in the sidecar container can access processes in the target container).

Reviewer 3: I don't now if it's worth calling out, but you can definitely use the diagnostics tools without sharing process namespace. I'm okay leaving this as is for now, and updating it in a future iteration once the tool can successfully consume the new process info commands.

@josalem Thoughts?

Reviewer: The socket discovery mechanics fall apart if you aren't sharing a PID-namespace and there are multiple container's apps you want to connect to, e.g., I have containers A, B, and C (my diag container).  If you don't share a process namespace, but you share `/tmp`, you'll end up multiple sockets that only differentiate by their epoch timestamp.  True we could use the process info command, but that only exists in 5.0 and there is a better mechanism than sharing `/tmp` in 5.0.

If you only have 1 container you want to connect to, then not sharing a PID-space is okay.  I think this nuance should be called out in the text if we remove the suggestion to share a PID-space.

These docs should include instructions that are limited to "best experience" for 3.1 I think.  We can add 5.0 specific instructions once that ships and the tools have finished being updated to use the new features.

Author: In the usual use case, won't customers need to share PID-space so that the `-p` parameters will work? If there's a reasonably straight-forward way to trace/watch/dump from another container without sharing process namespace, I can include that, but it might be better to wait and update in the .NET 5 timeframe with the updated experience.

> 2. The .NET Core global CLR diagnostic tools need access to files the .NET Core runtime writes to the /tmp directory, so the /tmp directory must be shared between the target and sidecar container via a volume mount. If you attempt to use the diagnostic tools from a sidecar container without sharing the /tmp directory, you will get an error about the process "not running compatible .NET runtime."

Reviewer 4: Forcing all CLR containers to mount a common `/tmp` directory seems messy/risky.  I could guess this is the side car mounting the CLR container's `/tmp` directory, but that seems to imply a nested container.  If that is the case the description is not very clear.

Sample sessions/scripts might be in order here and for the samples below.

Reviewer: I think a simple k8s manifest or pair of docker cli commands would work here.  For example, something like

```
docker run --rm -v scratch:/tmp appcontainer
```

```
docker run -it -v scratch:/tmp sidecar bash
```

would probably be sufficient.

Reviewer 3: Nit: you'll need to create the volume using
```
docker volume create scratch
```

Reviewer: If this is a docker server running on my dev box and `scratch` is a local directory, wouldn't the above be sufficient?

Reviewer 3: Assuming you have a local directory by that name and you've granted the docker daemon access to that volume.
Since the domain socket is ephemeral, there's not much value in sharing it with the host.

I found creating a volume is just more reliable for this scenario

Author: I'll add Docker commands or K8s manifest snippets to try make this clearer.

## Using `PerfCollect` in a container

The [`PerfCollect`](https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/linux-performance-tracing.md) script is useful for collecting performance traces and is the recommended tool for collecting traces prior to .NET Core 3.0. If using `PerfCollect` in a container, keep the following requirements in mind:

1. `PerfCollect` requires the [`SYS_ADMIN` capability](https://man7.org/linux/man-pages/man7/capabilities.7.html) (in order to run the `perf` tool), so be sure the container is [started with that capability](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities).

> 2. `PerfCollect` requires that the following environment variables be set prior to the app it is profiling starting. Because Docker entrypoints start almost immediately, it's necessary to define these environment variables when starting the container.

Reviewer 3: I found this unnecessarily confusing. I'd rather see a sample Dockerfile (like in the case of installing the CLI tools) that sets these variables

Author: These may not be in a Dockerfile, though, because customers may set the env vars in a K8s manifest or in some other way so that they're defined when starting the container. I'll try to rephrase this section to make it clearer.

    1. COMPlus_PerfMapEnabled=1
    1. COMPlus_EnableEventLog=1

### Using `PerfCollect` in a sidecar container

If you would like to run `PerfCollect` in one container to profile a .NET Core process in a different container, the experience is almost the same except for these differences:

1. The environment variables mentioned previously (COMPlus_PerfMapEnabled and COMPlus_EnableEventLog) must be set for the target container (not the one running `PerfCollect`).
2. The container running `PerfCollect` must have the `SYS_ADMIN` capability (not the target container).

> 3. Of course, the two containers must [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid).

Reviewer 3: nit:
```suggestion
3. The two containers must [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid).
```

Author: Fixed.

## Using `createdump` in a container

> An alternative to `dotnet-dump`, [`createdump`](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/xplat-minidump-generation.md) is the recommended tool for creating core dumps on Linux containing both native and managed information. The `createdump` tool is installed with the .NET Core runtime and can be found next to libcoreclr.so (typically in "/usr/share/dotnet/shared/Microsoft.NETCore.App/[version]"). The tool works the same in a container as it does in non-containerized Linux environments with the single exception that the tool requires the [`SYS_PTRACE` capability](https://man7.org/linux/man-pages/man7/capabilities.7.html), so the Docker container must be [started with that capability](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities).

Author: I meant to say that `createdump` is the preferred tool if just seeing managed stacks/heap/etc. isn't sufficient. If someone's only looking at managed code, I think I'd actually start with recommending dotnet-dump. That may not have been clear, though. I can rephrase as you suggest to make sure people don't think createdump should be used over dotnet-dump.

> An alternative to `dotnet-dump`, [`createdump`](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/xplat-minidump-generation.md) is the recommended tool for creating core dumps on Linux containing both native and managed information. The `createdump` tool is installed with the .NET Core runtime and can be found next to libcoreclr.so (typically in "/usr/share/dotnet/shared/Microsoft.NETCore.App/[version]"). The tool works the same in a container as it does in non-containerized Linux environments with the single exception that the tool requires the [`SYS_PTRACE` capability](https://man7.org/linux/man-pages/man7/capabilities.7.html), so the Docker container must be [started with that capability](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities).

Reviewer 4: Under the covers, dotnet-dump uses `createdump` to create a dump, so I don't think there is a benefit to preferring `createdump`.  

`dotnet-dump` runs `createdump` from the CLR process, so it has the advantage of not needing `SYS_ptrace` capability.

> An alternative to `dotnet-dump`, [`createdump`](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/xplat-minidump-generation.md) is the recommended tool for creating core dumps on Linux containing both native and managed information. The `createdump` tool is installed with the .NET Core runtime and can be found next to libcoreclr.so (typically in "/usr/share/dotnet/shared/Microsoft.NETCore.App/[version]"). The tool works the same in a container as it does in non-containerized Linux environments with the single exception that the tool requires the [`SYS_PTRACE` capability](https://man7.org/linux/man-pages/man7/capabilities.7.html), so the Docker container must be [started with that capability](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities).

Author: Ah, good to know. Thanks for clarifying, @sdmaclea. I've updated the doc to not call either tool 'preferred'.

### Using `createdump` in a sidecar container

If you would like to use `createdump` to create a dump from a process in a different container, the experience is almost the same except for these differences:

1. The container running `createdump` must have the `SYS_PTRACE` capability (not the target container).

> 2. Of course, the two containers must [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid).

Reviewer 3: nit:
```suggestion
2. The two containers must [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid).
```

Author: Fixed

### Using .NET Core global CLI tools in a sidecar container

If you would like to use .NET Core global CLI diagnostic tools to diagnose processes in a different container, bear the following additional requirements in mind:

1. The containers must [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid) (so that tools in the sidecar container can access processes in the target container).
2. The .NET Core global CLI diagnostic tools need access to files the .NET Core runtime writes to the /tmp directory, so the /tmp directory must be shared between the target and sidecar container via a volume mount. This could be done, for example, by having the containers share a common [volume](https://docs.docker.com/storage/volumes/#create-and-manage-volumes) or a Kubernetes [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume. If you attempt to use the diagnostic tools from a sidecar container without sharing the /tmp directory, you will get an error about the process "not running compatible .NET runtime."

## Using `PerfCollect` in a container

> **This article applies to: ✔️** .NET Core 2.1 and later versions

Reviewer 2: I think you missed one
```suggestion
**This tool applies to: ✔️** .NET Core 2.1 and later versions
```

Author: Good catch; fixed!
